The task was completed according to the instructions

zaitsev-serhei opened the pull request:
Created implementations of given interfaces and created classes according to the task.

zaitsev-serhei committed:
The task was completed following the instructions in the Maven
Created implementations of given interfaces and created classes according to the task.

Rommelua commented on src/main/java/core/mate/academy/coreelements/MachineGun.java:
```diff
@@ -0,0 +1,31 @@
+
+public class MachineGun {
+    private String name;
+    private int numberofShots;
```
```suggestion
    private int numberOfShots;
```
same for set/get

Rommelua commented on src/main/java/core/mate/academy/service/BulldozerProducer.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.List;
+
+public class BulldozerProducer implements MachineProducer<Machine> {
+    private List<Machine> bulldozerList;
```
This List should be a method local variable.

Rommelua commented on src/main/java/core/mate/academy/service/BulldozerProducer.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class BulldozerProducer implements MachineProducer<Machine> {
```
```suggestion
public class BulldozerProducer implements MachineProducer<Bulldozer> {
```

Rommelua commented on src/main/java/core/mate/academy/service/ExcavatorProducer.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.ArrayList;
+import java.util.List;
+
+class ExcavatorProducer implements MachineProducer<Machine> {
```
```suggestion
public class ExcavatorProducer implements MachineProducer<Excavator> {
```

Rommelua commented on src/main/java/core/mate/academy/service/TruckProducer.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.ArrayList;
+import java.util.List;
+
+class TruckProducer implements MachineProducer<Machine> {
```
```suggestion
public class TruckProducer implements MachineProducer<Truck> {
```

Rommelua commented on src/main/java/core/mate/academy/service/MachineServiceImpl.java:
```diff
@@ -1,7 +1,72 @@
+    }
+
+    @Override
+    public void fill(List<? super Machine> machines, Machine value) {
```
The task is _Fill the machines list with passed value_
You should set **value** instead of all objects in the List.

Rommelua commented on src/main/java/core/mate/academy/service/MachineServiceImpl.java:
```diff
@@ -1,7 +1,72 @@
+        }
+    }
+
+    public void fillBulldozer(List<? super Machine> machines) {
+        Bulldozer insertElem = new Bulldozer();
+        for (int i = 0; i < machines.size(); i++) {
+            machines.set(i,insertElem);
+        }
+    }
+
+    public void fillTruck(List<? super Machine> machines) {
+        Truck insertElem = new Truck();
+        for (int i = 0; i < machines.size(); i++) {
+            machines.set(i,insertElem);
+        }
+    }
+
+    public void fillExcavator(List<? super Machine> machines) {
+        Excavator insertElem = new Excavator();
+        for (int i = 0; i < machines.size(); i++) {
+            machines.set(i,insertElem);
+        }
+    }
```
```suggestion

```

Rommelua commented on src/main/java/core/mate/academy/service/MachineServiceImpl.java:
```diff
@@ -1,7 +1,72 @@
+
+    @Override
+    public void startWorking(List<?extends Machine> list) {
+        for (int i = 0; i < list.size(); i++) {
+            list.get(i).doWork();
+        }
```
better use foreach loop or forEach() method

Rommelua reviewed (changes requested):
coreelements is part of models. So, package structure should be:
.../model/elements
For ExcavatorBucket (if it is for Excavator only) possible solution to make it an inner class. But it is out of scope for this task - it's aboute generics.

zaitsev-serhei committed:
Updated package location for coreelements package
updated classes as in PR comments.
updated interface MachineProducer.java to apply changes from PR comments

Rommelua commented on src/main/java/core/mate/academy/service/MachineProducer.java:
```diff
@@ -1,11 +1,12 @@
      * @return - the list of models
      */
-    List<Object> get();
+    List<? super T> get();
```
```suggestion
    List<T> get();
```
It was correct before, why you changed it?

zaitsev-serhei committed:
updated MachineServiceImpl method fill.
updated MachineProducer.java and its implementations to return the supposed type

Rommelua commented on src/main/java/core/mate/academy/service/MachineServiceImpl.java:
```diff
@@ -1,7 +1,46 @@
+    @Override
+    public List<Machine> getAll(Class<? extends Machine> type) {
+        if (type.equals(Bulldozer.class)) {
+            List<? extends Machine> machines = new BulldozerProducer().get();
```
```suggestion
            List<Bulldozer> machines = new BulldozerProducer().get();
```

zaitsev-serhei committed:
updated the getAll method in MachineServiceImpl
